### ISSUE TITLE
BUG: np.fft.hfft on main segfaults when n=1

### ISSUE THREAD

**Issue by jakevdp** [numpy/numpy#25661]
### Describe the issue:

With the current nightly release, `np.fft.hfft` segfaults when `n=1`

### Reproduce the code example:

```python
In [1]: import numpy as np

In [2]: np.__version__
Out[2]: '2.0.0.dev0+git20240120.6bd3abf'

In [3]: x = np.arange(10)

In [4]: np.fft.hfft(x, n=1)
Segmentation fault: 11
```

### Error message:

```shell
Segmentation fault: 11
```

### Python and NumPy Versions:

```python
>>> import sys, numpy; print(numpy.__version__); print(sys.version)
2.0.0.dev0+git20240120.6bd3abf
3.9.6 (default, May  7 2023, 23:32:44) 
[Clang 14.0.3 (clang-1403.0.22.14.1)]
```

### Runtime Environment:

```python
>>> import numpy; print(numpy.show_runtime())
[{'numpy_version': '2.0.0.dev0+git20240120.6bd3abf',
  'python': '3.9.6 (default, May  7 2023, 23:32:44) \n'
            '[Clang 14.0.3 (clang-1403.0.22.14.1)]',
  'uname': uname_result(system='Darwin', node='jmdg-macbookpro.roam.internal', release='23.2.0', version='Darwin Kernel Version 23.2.0: Wed Nov 15 21:53:18 PST 2023; root:xnu-10002.61.3~2/RELEASE_ARM64_T6000', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}}]
```

### Context for the issue:

I found this through nightly tests of JAX against upstream NumPy; this failure began to appear in the last few days, so this is likely due to a recent change to `fft`.

**Comment by jakevdp** [comment]
I haven't bisected, but given the timing #25536 seems to be the likely culprit. cc/ @mhvk

**Comment by ngoldbaum** [comment]
Seems to be coming from `memcpy` being passed a negative size:

```
(lldb) up
warning: _pocketfft_umath.cpython-311-darwin.so was compiled with optimization - stepping may behave oddly; variables may not be available.
frame #1: 0x00000001003e6990 _pocketfft_umath.cpython-311-darwin.so`irfft_loop [inlined] copy_data(in=<unavailable>, step_in=16, nin=8, out="", step_out=16, nout=-1, elsize=16) at _pocketfft_umath.c:37:9 [opt]
   34  	{
   35  	    npy_intp ncopy = nin <= nout? nin : nout;
   36  	    if (step_in == elsize && step_out == elsize) {
-> 37  	        memcpy(out, in, ncopy*elsize);
   38  	    }
   39  	    else {
   40  	        char *ip = in, *op = out;
(lldb) p ncopy
(npy_intp) -1
(lldb) p nin
(npy_intp) 8
(lldb) p nout
(npy_intp) -1
```

**Comment by mhvk** [comment]
That surely is me :facepalm: 
...
Fix in #25668.
